### PR TITLE
Tests for logger

### PR DESCRIPTION
**Pull request by waynevanson**
#80 Tests for the logger.

I've used `jest.resetModules()` alongside `beforeEach()` and `afterEach()` to change `process.env.NODE_ENV` from `"test"` to `"production"`, as `logger[name]` relies on `NODE_ENV` not to be in test mode.

Test coverage is 100% for this module.

~~Can I have someone check out these tests? I've never written tests like this before and am unsure what the standard.~~

~~I made the helper functions because I'll use them for the other log functions.~~

~~I extended this branch off a previous pull, which is why there is all this other stuff. I'll get rid of it later before opening up.~~

### CODE REVIEW

**Comment by sQVe** [comment]
I do not think it is a good idea to manipulate `NODE_ENV`.

If we refactor the individual methods in `logger` like this:
```typescript
export const info = (msg: string) => {
  const text = chalk`{green.bold INFO:} ${msg}`;

  if (process.env.NODE_ENV !== "test") {
    console.log(text);
  }

  return text;
};
```
We could instead just test the return value easily, either manually or via a snapshot.

**Comment by waynevanson** [comment]
I like that a lot better! I'll add that in.

**Comment by waynevanson** [comment]
@sQVe If we are using NODE_ENV as a conditional, we can't test the case outside of test without changing it.

I think as long as we're mindful that we're changing it, it's okay. Otherwise we won't have the code coverage for it. All we'd be checking for is the colour of the text, which is half it's function.

What about something like this? It changes the env, calls the function then sets the previous environment back.
```ts
function useNodeEnv(
  env: "development" | "test" | "production",
  handler: () => void
) {
  const defaultEnv = process.env.NODE_ENV.toString();
  process.env.NODE_ENV = env;
  handler();
  process.env.NODE_ENV = defaultEnv;
}

useNodeEnv('production',() => {/* test in environment */})
```
```

**Comment by sQVe** [comment]
I think it is fine that the `NODE_ENV === 'test'` are not covered and can be excluded from coverage. If your going to mutate `NODE_ENV` that should be done in a `beforeEach` and `afterEach` with proper cleanup via `jest.resetModules()` or something similar.

**Comment by waynevanson** [comment]
Okay, I think I can do that after reading some stuff online. Thank you for the tips!

File: src/shared/logger.ts [hunk 1]
```diff
@@ -2,21 +2,18 @@ import chalk from "chalk";
...
     "If you think this is a bug, you can report it: https://github.com/benawad/destiny/issues"
   );
-
   process.exit(code);
 };
 
 export const info = (msg: string) => {
   if (process.env.NODE_ENV === "test") return;
-  console.info(chalk`{green.bold INFO:} ${msg}`);
+  const text = chalk.green.bold(`INFO: ${msg}`);
```
Comment (AnatoleLucet) [inline src/shared/logger.ts]: Is this wanted that it's logging the full message in green and not only the prefix `INFO:` (same for the `warn`)?
I think this is a bit too much...

Comment (waynevanson) [reply src/shared/logger.ts]: I see what you mean. It's definitely too much, I'll change it now.

File: tests/logger.test.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+describe('NODE_ENV === "test"', () => {
+  describe.each(Object.keys(logger))("logger.%s", name => {
+    test(`stops execution of logger.${name} if in a test environment`, () => {
+      logger[name]("a test message");
+      expect(mocks.error).toBeCalledTimes(0);
+      expect(mocks.info).toBeCalledTimes(0);
+      expect(mocks.log).toBeCalledTimes(0);
+      expect(mocks.warn).toBeCalledTimes(0);
+      expect(mocks.exit).toBeCalledTimes(0);
```
Comment (AnatoleLucet) [inline tests/logger.test.ts]: What about: ?
```suggestion
      for (const mock of Object.values(mocks)) {
        expect(mock).toBeCalledTimes(0);
      }
```

File: tests/logger.test.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      logger[name](message);
+      const mock = mocks[name];
+      expect(mock).toBeCalledTimes(1);
+      expect(mock).toBeCalledWith(msg);
+    });
+  });
+
+  describe(`logger.error`, () => {
+    it('calls "console.error()" when passed an error instances', () => {
+      const error = new Error();
```
Comment (AnatoleLucet) [inline tests/logger.test.ts]:
```suggestion
      const error = new Error("test error");
```
So we are sure it logs the given error and not just a random one.

File: tests/logger.test.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      expect(mocks.error).toBeCalledTimes(1);
+      expect(mocks.error).toBeCalledWith(result);
+      expect(mocks.exit).toBeCalledTimes(1);
+    });
+
+    it('exits process with error code "1"', () => {
+      logger.error("", 1);
+      expect(mocks.error).toBeCalledTimes(1);
+      expect(mocks.exit).toBeCalledTimes(1);
+      expect(mocks.exit).toBeCalledWith(1);
```
Comment (AnatoleLucet) [inline tests/logger.test.ts]: It's just a bit clearer.
```suggestion
      const exitCode = 1;
      logger.error("", exitCode);
      expect(mocks.error).toBeCalledTimes(1);
      expect(mocks.exit).toBeCalledTimes(1);
      expect(mocks.exit).toBeCalledWith(exitCode);
```

**Pull request merged**